File: geon-js/examples/07_file_io.ts
import { GeonPlace, generate, parse } from '../src';
import * as fs from 'fs';
import * as path from 'path';

// 1. Create a sample .geon object
const place = new GeonPlace();
place.place = "Victoria Park, Nottingham";
place.type = "public_space";
place.location = { lat: 52.9403, lon: -1.1340 };
place.source = ["OpenStreetMap (2025-01)"];

// Write to a temp file
const filename = path.join(__dirname, "example_07_temp.geon");
const text = generate(place);

fs.writeFileSync(filename, text, 'utf8');
console.log(`Wrote GEON file: ${filename}`);
console.log(`Size: ${text.length} bytes\n`);

// 2. Read it back
const loadedText = fs.readFileSync(filename, 'utf8');
const loaded = parse(loadedText);

console.log(`Loaded: ${loaded.place}`);
console.log(`Type:   ${loaded.type}`);
console.log();

// 3. Modify and save
loaded.temporal["summer_events"] = "concerts every Friday July-August";
loaded.adjacencies.push("Sneinton Market (500m north)");

const updatedText = generate(loaded);
const filename2 = path.join(__dirname, "example_07_temp_updated.geon");
fs.writeFileSync(filename2, updatedText, 'utf8');

console.log(`Updated GEON file: ${filename2}`);
console.log(`Size: ${updatedText.length} bytes\n`);
console.log("=== Updated content ===");
console.log(updatedText);

// Clean up
try {
    fs.unlinkSync(filename);
    fs.unlinkSync(filename2);
} catch (e) {
    // ignore
}

File: geon-rs/examples/07_file_io.rs
use geon_rs::{GeonPlace, Coordinate, generate, parse};
use std::collections::HashMap;
use std::fs::File;
use std::io::{Read, Write};

fn main() -> std::io::Result<()> {
    // 1. Create a sample .geon file in memory
    let mut place = GeonPlace::default();
    place.place = "Victoria Park, Nottingham".to_string();
    place.type_ = "public_space".to_string();
    place.location = Some(Coordinate::new(52.9403, -1.1340));
    place.area = Some("14 hectares".to_string());
    place.purpose = vec!["recreation".to_string(), "sport".to_string(), "ecology".to_string(), "events".to_string()];
    
    let mut exp = HashMap::new();
    exp.insert("openness".to_string(), "high".to_string());
    exp.insert("enclosure".to_string(), "low".to_string());
    exp.insert("activity_density".to_string(), "moderate".to_string());
    place.experience = exp;
    
    place.character = vec![
        "Victorian (established 1880s)".to_string(),
        "well-maintained (bowling greens, flower beds)".to_string(),
    ];
    place.source = vec!["OpenStreetMap (2025-01)".to_string()];

    // Write to a temp file
    // In Rust std::env::temp_dir() or just local file for example
    let filename = "example_07_temp.geon";
    let text = generate(&place);
    
    {
        let mut file = File::create(filename)?;
        file.write_all(text.as_bytes())?;
    }
    
    println!("Wrote GEON file: {}", filename);
    println!("Size: {} bytes\n", text.len());
    
    // 2. Read it back
    let loaded_text = std::fs::read_to_string(filename)?;
    let mut loaded = parse(&loaded_text);
    
    println!("Loaded: {}", loaded.place);
    println!("Type:   {}", loaded.type_);
    println!("Valid:  true (implied)");
    println!();
    
    // 3. Modify and save
    let mut temp = HashMap::new();
    temp.insert("summer_events".to_string(), "concerts every Friday July-August".to_string());
    temp.insert("parkrun".to_string(), "every Saturday 09:00".to_string());
    loaded.temporal = temp;
    
    loaded.adjacencies = vec![
        "Sneinton Market (500m north)".to_string(),
        "Nottingham Station (1.2km west)".to_string(),
    ];
    
    let updated_text = generate(&loaded);
    let filename2 = "example_07_temp_updated.geon";
    {
        let mut file = File::create(filename2)?;
        file.write_all(updated_text.as_bytes())?;
    }

    println!("Updated GEON file: {}", filename2);
    println!("Size: {} bytes\n", updated_text.len());
    println!("=== Updated content ===");
    println!("{}", updated_text);
    
    // Clean up
    std::fs::remove_file(filename)?;
    std::fs::remove_file(filename2)?;

    Ok(())
}
